### PR TITLE
True half-precision support in Fabric

### PR DESCRIPTION
**Pull request by awaelchli**
Adds support for half-precision training and inference with Fabric:

```py
fabric = Fabric(precision="16-true")
fabric = Fabric(precision="bf16-true")

model = ...
model = fabric.setup(model)  # model gets converted to the half type
```

To save peak memory usage, you can init the model directly with the desired precision like so:

```py
fabric = Fabric(precision="bf16-true")

with fabric.init_module():
    model = ...

print(model.dtype)
```

For training, there will most likely be instabilities when choosing precision="16-true", so mixed precision training is still preferred in many cases. This PR offers more flexibility for inference, especially with bfloat16. See Lit-LLaMA for example: https://github.com/Lightning-AI/lit-llama

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/connector.py [hunk 1]
```diff
@@ -468,8 +471,8 @@ def _check_and_init_precision(self) -> Precision:
             device = "cpu" if self._accelerator_flag == "cpu" else "cuda"
 
             if isinstance(self.strategy, FSDPStrategy):
-                return FSDPPrecision(precision=self._precision_input, device=device)
-            return MixedPrecision(precision=self._precision_input, device=device)
+                return FSDPPrecision(precision=self._precision_input, device=device)  # type: ignore[arg-type]
+            return MixedPrecision(precision=self._precision_input, device=device)  # type: ignore[arg-type]
```
Comment (awaelchli) [inline src/lightning/fabric/connector.py:475]: weird mypy thinks we are passing in "16-true" even though above we check `self._precision_input in ("16-mixed", "bf16-mixed")`

**Review by Borda — APPROVED** [review]

File: docs/source-fabric/fundamentals/precision.rst [hunk 2]
```diff
@@ -118,6 +118,41 @@ It is also possible to use BFloat16 mixed precision on the CPU, relying on MKLDN
...
+    fabric = Fabric(precision="16-true")
+    model = MyModel()
+    model = fabric.setup(model)  # model gets cast to torch.float16
+
+    # Select BF16 precision
+    fabric = Fabric(precision="bf16-true")
+    model = MyModel()
+    model = fabric.setup(model)  # model gets cast to torch.bfloat16
+
+Tipp: For faster initialization, you can create model parameters with the desired dtype directly on the device:
```
Comment (carmocca) [inline docs/source-fabric/fundamentals/precision.rst:140]:
```suggestion
Tip: For faster initialization, you can create model parameters with the desired dtype directly on the device:
```

File: src/lightning/fabric/plugins/precision/half.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from lightning_utilities.core.apply_func import apply_to_collection
+from torch import Tensor
+from torch.nn import Module
+
+from lightning.fabric.plugins.precision.precision import Precision
+from lightning.fabric.plugins.precision.utils import _convert_fp_tensor
+
+
+class HalfPrecision(Precision):
+    """Plugin for training with half (``torch.bfloat16``) precision.
```
Comment (carmocca) [inline src/lightning/fabric/plugins/precision/half.py:27]:
```suggestion
    """Plugin for training with half precision.
```

File: src/lightning/fabric/plugins/precision/half.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    Args:
+        precision: Whether to use ``torch.float16`` (``'16-true'``) or ``torch.bfloat16`` (``'bf16-true'``).
+    """
+
+    precision: Literal["bf16-true", "16-true"] = "bf16-true"
+
+    def __init__(self, precision: Literal["bf16-true", "16-true"]) -> None:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/fabric/plugins/precision/half.py:35]: The precision class attribute already sets a default. Either have the same default or no default

File: src/lightning/fabric/plugins/precision/half.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        self.precision = precision
+        self._desired_input_dtype = torch.bfloat16 if precision == "bf16-true" else torch.float16
+
+    def convert_module(self, module: Module) -> Module:
+        return module.bfloat16() if self.precision == "bf16-true" else module.half()
```
Comment (carmocca) [inline src/lightning/fabric/plugins/precision/half.py:40]:
```suggestion
        return module.to(dtype=self._desired_input_dtype)
```

File: tests/tests_fabric/strategies/test_ddp.py [hunk 6]
```diff
@@ -133,6 +133,8 @@ def test_ddp_grad_clipping(clip_type, accelerator, precision):
     "precision,expected_dtype",
     [
         (Precision(), torch.float32),
+        (HalfPrecision("16-true"), torch.float16),
+        pytest.param(HalfPrecision("bf16-true"), torch.bfloat16, marks=RunIf(bf16_cuda=True)),
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_ddp.py:137]: GPU CI only picks tests that set `min_cuda_gpus`. If you set just `bf16_cuda` then this is effectively skipped.

Could this run on CPU? If not, then you should add `min_cuda_gpus`

Same for the other tests

File: src/lightning/fabric/connector.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/connector.py:475]: Probably a bug in mypy

File: tests/tests_fabric/strategies/test_ddp.py (hunk 6 above)
Comment (awaelchli) [reply tests/tests_fabric/strategies/test_ddp.py:137]: It already has @RunIf(min_cuda_gpus=2) at the top, and there are other tests where it's on the CPU

**Review by carmocca — APPROVED** [review]

**Pull request merged**